### PR TITLE
[ARM] fix "+fp.dp" in multilib selection

### PR DESCRIPTION
**Pull request by domin144**
When the FPU was selected with "+(no)fp(.dp)" extensions in "-march" or "-mcpu" options, the FPU used for multilib selection was still the default one for given architecture or CPU.

### CODE REVIEW

File: llvm/lib/TargetParser/ARMTargetParser.cpp [hunk 1]
```diff
@@ -366,26 +366,50 @@ StringRef ARM::getArchExtFeature(StringRef ArchExt) {
...
+}
+
+static ARM::FPUKind findSinglePrecisionFPU(ARM::FPUKind InputFPUKind) {
+  if (InputFPUKind == ARM::FK_INVALID || InputFPUKind == ARM::FK_NONE)
+    return ARM::FK_INVALID;
+
   const ARM::FPUName &InputFPU = ARM::FPUNames[InputFPUKind];
 
   // If the input FPU already supports double-precision, then there
   // isn't any different FPU we can return here.
```
Comment (john-brawn-arm) [inline llvm/lib/TargetParser/ARMTargetParser.cpp:400]: This comment should be above in findDoublePrecisionFPU, the comment here should be the opposite (i.e. should say the input FPU already doesn't have double-precision).

File: llvm/lib/TargetParser/ARMTargetParser.cpp [hunk 2]
```diff
@@ -420,20 +444,35 @@ bool ARM::appendArchExtFeatures(StringRef CPU, ARM::ArchKind AK,
     CPU = "generic";
 
   if (ArchExt == "fp" || ArchExt == "fp.dp") {
+    const ARM::FPUKind DefaultFPU = getDefaultFPU(CPU, AK);
     ARM::FPUKind FPUKind;
     if (ArchExt == "fp.dp") {
+      const bool IsDP = ArgFPUKind != ARM::FK_INVALID &&
+                        ArgFPUKind != ARM::FK_NONE &&
+                        isDoublePrecision(getFPURestriction(ArgFPUKind));
```
Comment (john-brawn-arm) [inline llvm/lib/TargetParser/ARMTargetParser.cpp:452]: This function is called in order to set ArgFPUKind, so checking it here (and below) looks strange. I would expect it to always be FK_INVALID, is there any situation in which it's not?

File: llvm/lib/TargetParser/ARMTargetParser.cpp (hunk 1 above)
Comment (domin144) [reply llvm/lib/TargetParser/ARMTargetParser.cpp:400]: Indeed. Thanks for spotting this.

File: llvm/lib/TargetParser/ARMTargetParser.cpp (hunk 2 above)
Comment (domin144) [reply llvm/lib/TargetParser/ARMTargetParser.cpp:452]: The `appendArchExtFeatures` is called repeatedly in `DecodeARMFeatures`. There can be multiple FPU related extensions specified. e.g. `-march=armv8-m.main+nofp+fp+nofp.dp` should disable FPU, enable default FPU(fpv5-d16) and then switch to single precission equivalent FPU(fpv5-sp-d16).

Comment (domin144) [reply llvm/lib/TargetParser/ARMTargetParser.cpp:452]: I will add a test case for this.

**Review by john-brawn-arm — APPROVED** [review]
Looks good to me.

**Pull request merged**